cmd/security-secrets-setup/main.go

```diff
@@ -16,80 +17,98 @@ package main
+	callExit(int)
+}
+
+type exitCode struct{}
+
+type optionDispatcher interface {
+	run() (int, error)
+}
+
+type pkiInitOptionDispatcher struct{}
+
+var exitInstance = newExit()
+var dispatcherInstance = newOptionDispatcher()
+var configFile string
 
+func init() {
+	// define and register command line flags:
 	flag.StringVar(&configFile, "config", "", "specify JSON configuration file: /path/to/file.json")
 	flag.StringVar(&configFile, "c", "", "specify JSON configuration file: /path/to/file.json")
 	flag.Usage = usage.HelpCallbackSecuritySetup
-	flag.Parse()
-
-	if configFile == "" {
-		flag.PrintDefaults()
-		os.Exit(1)
-	}
+}
 
-	setup.Init()
+func main() {
+	start := time.Now()
 
-	// Create and initialize the fs environment and global vars for the PKI materials
-	lc := logger.NewClient("security-secrets-setup", setup.Configuration.Logging.EnableRemote,
-		setup.Configuration.Logging.File, setup.Configuration.Writable.LogLevel)
+	flag.Parse()
 
-	// Read the Json config file and unmarshall content into struct type X509Config
-	x509config, err := config.NewX509Config(configFile)
-	if err != nil {
-		lc.Error(err.Error())
+	if len(os.Args) < 2 {
+		fmt.Println("Please specify option for " + option.SecuritySecretsSetup)
+		flag.PrintDefaults()
+		exitInstance.callExit(1)
 		return
 	}
 
-	seed, err := setup.NewCertificateSeed(x509config, setup.NewDirectoryHandler(lc))
-	if err != nil {
-		lc.Error(err.Error())
+	if err := setup.Init(); err != nil {
+		exitInstance.callExit(1)
 		return
 	}
 
-	rootCA, err := certificates.NewCertificateGenerator(certificates.RootCertificate, seed, certificates.NewFileWriter(), lc)
-	if err != nil {
-		lc.Error(err.Error())
-		return
-	}
+	if configFile == "" {
+		// run with other options for pki-init
+		statusCode, err := dispatcherInstance.run()
+		if err != nil {
+			setup.LoggingClient.Error(err.Error())
+		}
 
-	err = rootCA.Generate()
-	if err != nil {
-		lc.Error(err.Error())
+		exitInstance.callExit(statusCode)
 		return
 	}
 
-	tlsCert, err := certificates.NewCertificateGenerator(certificates.TLSCertificate, seed, certificates.NewFileWriter(), lc)
-	if err != nil {
-		lc.Error(err.Error())
+	if err := option.GenTLSAssets(configFile); err != nil {
+		setup.LoggingClient.Error(err.Error())
 		return
 	}
 
-	tlsCert.Generate()
-	if err != nil {
-		lc.Error(err.Error())
-		return
-	}
-	lc.Info("PKISetup complete", internal.LogDurationKey, time.Since(start).String())
+	setup.LoggingClient.Info(option.SecuritySecretsSetup+" complete", internal.LogDurationKey, time.Since(start).String())
+}
+
+func newExit() exit {
+	return &exitCode{}
+}
+
+func (code *exitCode) callExit(statusCode int) {
+	os.Exit(statusCode)
+}
+
+func newOptionDispatcher() optionDispatcher {
+	return &pkiInitOptionDispatcher{}
 }
 
-// TODO: ELIMINATE THIS ----------------------------------------------------------
-func fatalIfErr(err error, msg string) {
+func setupPkiInitOption() (executor option.OptionsExecutor, status int, err error) {
```

Comment:
this function doesn't do much and is only used inside the `run()` method so I think you should just roll it into `run()`. Where you use it in the tests can be similarly rolled inside, running a single function with a default value struct doesn't really warrant sharing code like this

Reply:
Well, this way is more unit-testable. So someone can just unit-test this setup to make usre it does what it needs to do and not worry about the driver part (which is `run()`).  As a matter of fact, in the unit tests, `run()` is overridden and used for test cases.

Reply:
this specific function is not unit tested though, `run()` is what you are unit testing, this function doesn't do anything except call one other function

Reply:
Adding unit test to it

Reply:
sorry the point I was making is that you don't need to unit test this function because it isn't doing much of anything, it's like defining this function:

```go
func doAThing() {
    return doAnotherthing(struct{})
}
```

there's nothing gained by this function and so it doesn't need a test and the function doesn't need to exist is all I'm saying, apologies if that was unclear

Reply:
Ian, there is something being doing in line 107 `option.NewPkiInitOption(opts)`, this function is doing something.  As a matter of fact, this one serves as the purpose of regression-testing later on when `-generate`, `-cache`, and `-import` option are going to be added.  With this as separate function, then it can be regression tested by the unit test to see if anything break when options are added later.